### PR TITLE
(bugfix): `gh-r` choosing wrong binary for OS

### LINKED ISSUE #160: 🐞 bug: bick default selects darwin when using manjaro
### Issue description

When "bpick" is not given explicitly the automatic selection of a release-binary selects darwin instead of linux.
```
❯ uname -a
Linux ithaka 5.13.19-2-MANJARO #1 SMP PREEMPT Sun Sep 19 21:31:53 UTC 2021 x86_64 GNU/Linux
```
For example [k9s](https://github.com/derailed/k9s/releases) is configured:
```
zcommand from"gh-r"; zload derailed/k9s
```

Installing / zinit update reports:
```
Downloading derailed/k9s…
(Requesting `k9s_Darwin_x86_64.tar.gz'…)
######################################################################## 100,0% 
ziextract: Unpacking the files from: `k9s_Darwin_x86_64.tar.gz'…
ziextract: Successfully extracted and assigned +x chmod to the file: `k9s'.
```

### zinit config

to much to be reported

### zinit version or commit ID

4da69625bcb7c12f929db84af2352e046eb11d75

### zsh version

5.8

### host info

XDG_SESSION_TYPE=x11 OSTYPE=linux-gnu CPUTYPE=x86_64 MACHTYPE=x86_64 5.13.19-2-MANJARO ID=manjaro

### PR DESCRIPTION
**Pull request by vladdoster**
- fix #160
- fix zinit selecting Linux version on macOS
- fix zinit not selecting `musl` version on `musl` systems
- add `gh-r` ice unit tests on Ubuntu & macOS to verify behavior on
  different OSes

Signed-off-by: Vladislav Doster <mvdoster@gmail.com>

### CODE REVIEW

**Comment by alichtman** [comment]
I don't have time to actually review this (or come up to speed) but I suggest reverting the at-fault diff ASAP, and sorting out the proper behavior without bugs on master. Not sure if this has been done already (on mobile)

**Comment by vladdoster** [comment]
@alichtman This sorts it out and adds a unit test to confirm that it does

**Comment by alichtman** [comment]
If @pschmitt doesn't have time to review it soon (within a day or two), I'll revert the at-fault diff until I have time to review it

**Comment by alichtman** [comment]
Can you make all the tests pass?

**Comment by alichtman** [comment]
And how does the test that was added guarantee that the right binary is downloaded for the arch it's running on? It seems like an incomplete test (or very targeted for our specific test env + this bug)

**Comment by vladdoster** [comment]
> If @pschmitt doesn't have time to review it soon (within a day or two), I'll revert the at-fault diff until I have time to review it.
> 
> I'm not sure exactly what happened (and I don't really care who committed what broken code). But we "can't break userspace"

I agree, but to be fair there were packages on Darwin that selected Linux before the commit.

Spot on -- it is targeted to Linux atm. I'll update the test.yml to be an OS matrix to be {Ubuntu,macOS} and update the test to account for that.

**Comment by vladdoster** [comment]
> Can you make all the tests pass?

The only test failing atm is `zshelldocs`, but `make doc` is adding + moving a lot of unrelated stuff around.

**Comment by vladdoster** [comment]
This can be merged but doesn't improve `darwin` or tricky packages like `shfmt`.

File: zinit-install.zsh [hunk 1]
```diff
@@ -1455,8 +1455,8 @@ builtin source "${ZINIT[BIN_DIR]}/zinit-side.zsh" || {
         amd64   "(x86_64|amd64|intel|linux64|linux_amd64)"
         aarch64 "aarch64"
         aarch64-2 "arm"
-        musl "*musl*"
-        darwin  "(darwin|mac|macos|osx|os-x)"
+        musl "*((#s)|/)*(linux|musl)*((#e)|/)*"
```
Comment (pschmitt) [inline zinit-install.zsh:1458]: Please add comments, the next person that reads this will be grateful. It might be you :)

Comment (vladdoster) [reply zinit-install.zsh:1458]: lmao, good catch

**Comment by vladdoster** [comment]
@pschmitt IMO this whole section for all OS/Archs could (and should) be reduced to gnu-{linux,musl} && macOS {intel,arm}. It's over-engineered but curious what you think.

**Comment by vladdoster** [comment]
I'll rebase my commits if you just want to get the fix out. I think I've been chasing a ghost in the tests to test for `musl` but confirming that now.

**Comment by vladdoster** [comment]
yep, so need to figure out a way to test for `linux-musl` without an explicit bpick'*musl*'. Checking GH docs

image

File: .github/workflows/tests.yaml [hunk 2]
```diff
@@ -12,6 +12,11 @@ jobs:
       - name: "💬 Set matrix output"
         id: set-matrix
         run: |
+          echo "${OSTYPE}"
+          echo "${OSTYPE}"
+          echo "${OSTYPE}"
+          echo "${OSTYPE}"
+          echo "${OSTYPE}"
```
Comment (jankatins) [inline .github/workflows/tests.yaml:19]: If this is intended, a comment might be good to explain why the duplication :-)

Comment (vladdoster) [reply .github/workflows/tests.yaml]: nope, I just rebased my commits, but IIRC the commit message said it was for testing.

Removed.

**Comment by vladdoster** [comment]
Closing in favor of #171

**Pull request closed without merging**